File: 백준/Silver/1051. 숫자 정사각형/숫자 정사각형.py
from sys import stdin
if __name__ =='__main__':
    
    n,m = (stdin.readline()).split(" ")
    n=int(n)
    m=int(m)
    _min=min(n,m)
    str_list=[""]*n
    
    for i in range(n):
        _in=stdin.readline()[:-1]
        str_list[i]=_in
    
    ans:int =1
    
    
    for i in range(n):
        for j in range(m):
            for l in range(1,_min):
                right:int=j+l
                bottom:int=i+l
                if(right>=m or bottom >=n): # 범위 벗어나면 다시 
                    break
                
                top_left=str_list[i][j]
                top_right=str_list[i][right]
                bottom_left=str_list[bottom][j]
                bottom_right=str_list[bottom][right]
                
                if(top_left==top_right and top_left==bottom_left and top_left==bottom_right): #모두 같은 값이면 
                    ans=max(ans,(l+1)**2) # 해당 (길이+1) 제곱, l=1일 때 해당 사각형 길이가 2 
    
    print(ans)
        
    
   

File: 백준/Silver/1051. 숫자 정사각형/숫자 정사각형.swift
import Foundation

func input() -> [Int] {
    return readLine()!.split{$0 == " "}.map{Int($0)!}
}

let nm = input()

let (n,m) = (nm[0],nm[1])

var board: [[Int]] = []

var result: Int = 1

for _ in 0..<n {
    board.append(Array(readLine()!).map{Int(String($0))!})
}

let limit = min(n,m)

func travel(_ x:Int,_ y:Int) {
    
    let target = board[x][y]
    
    
    for l in 1..<limit {
        let x2 = x+l
        let y2 = y+l
        
        if x2 >= n || y2 >= m {
            break
        }
        
        if target ==  board[x][y2] && target == board[x2][y] &&   target == board[x2][y2] {
            result = max(result,(l+1)*(l+1))
        }
        
    }
    
}

for i in 0..<n {
    for j in 0..<m {
    travel(i, j)
    }
}

print(result)
